### PR TITLE
Chained import → process → move-to-NAS workflow

### PR DESCRIPTION
**Pull request by jss367**
## What

Restores the retired local-processing workflow's key property — process on fast local disk, pay the network cost once, end on the NAS — as a chain of existing jobs configured once at import time:

**Import to local archive → process locally → each imported folder moves to the NAS automatically.**

- `local_archive_root` on remote targets maps the local archive to the NAS root (Settings → Remote targets); the move mirrors the layout (`<root>/2026/trip` → `<remote_path>/2026/trip`, catalog repoints to the mount)
- "Then move to NAS" option in the Import page's After-import card, with an honest preview of the exact landing path and warnings for every condition that would make the move fail (missing rsync/ssh, missing/relative mount path)
- New `after_process_move` request key, validated up front and snapshotted at enqueue (a Settings edit mid-chain can't redirect the move)
- Process-job completion hook enqueues merge-mode move-folder jobs: fires on process success **and failure** (unattended runs must still end on the NAS), skips only on explicit cancel — including cancel while a chained move waits its turn
- Transfers are serialized by a chain-local lock (one rsync to the NAS at a time, bwlimit honored); waiting jobs report "Waiting for an earlier chained move to finish"
- Move-folder guard stack and job start extracted into shared, job-thread-safe helpers reused by the endpoint and the chain
- Restart mid-chain fails safe: photos stay cataloged in the local archive; move manually via the Move page

Spec: `docs/superpowers/specs/2026-07-19-import-process-move-chain-design.md`
Plan: `docs/superpowers/plans/2026-07-19-import-process-move-chain.md`

## Tests

Full suite: **1767 passed, 14 skipped** (workspaces, db, app, photos/edits/jobs/darktable/config APIs, new `test_import_chain.py`).

New coverage: config coercion round-trip for `local_archive_root`; `minimal_move_set` unit tests (nesting, prefix-siblings, outside-root); endpoint validation matrix (7 tests); chain integration with stubbed move machinery — happy path (move jobs enqueue with merge + mount-side destinations + `chained_from`), process-raise path (move still fires), process-cancel path (no move, `after_move_skipped`), cancel-during-lock-wait (deterministic two-folder test).

Manual smoke against a live instance: Settings field round-trips through save/load; `/api/remote-targets` serves the field; all three request-validation rejections return honest 400s; Import/Settings pages render the new UI.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added an optional “Then move to NAS” chained step after archive imports and processing.
  * Added NAS move previews with eligibility gating and clearer Import-page warnings.
  * Introduced “Local archive root” for remote targets to enable chained move behavior.
* **Bug Fixes**
  * Improved chained-move reliability: moves still enqueue when processing fails, and skip behavior is now correct when explicitly cancelled.
  * Strengthened destination validation and corrected root-level edge-case “honest skip” reporting.
* **Documentation**
  * Added workflow documentation and design/specs for the Import → Process → Move-to-NAS chain.
* **Tests**
  * Added coverage for move-chain planning, validation rules, and concurrency cancellation scenarios.

### CODE REVIEW

File: vireo/import_chain.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    root = os.path.realpath(archive_root)
+    inside = []
+    for folder_id, path in folders:
+        real = os.path.realpath(path)
+        try:
+            common = os.path.commonpath([real, root])
+        except ValueError:
+            continue  # different drives — cannot be under the root
+        if common != root or real == root:
+            continue
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/import_chain.py]: Addressed in two layers (c24f975f + 071bc42f): the request-time guard in `_validate_after_process_move` rejects destination == `local_archive_root` when the folder template renders empty/`.` (destination == root with a dated template stays legal — that's the normal layout), and because no static check can see every template that renders to the root at runtime (`./.` normalizes into the destination; a strftime directive can render empty for some locales), `minimal_move_set` now returns `(moves, skipped)` with per-folder reasons and the chain carries an honest `skip_note`: the import result card says the move will not run (or that some photos stay local) and points at the Move page, and the process job records the same reason in `after_move_skipped`/`after_move_note` instead of a bare "no folders to move". Covered by `test_chain_root_level_import_reports_honest_skip` plus new `minimal_move_set` unit tests.

File: vireo/app.py [hunk 2]
```diff
@@ -20963,6 +21111,17 @@ def _fs_is_case_insensitive(path):
...
+        # Validate the optional NAS move that chains after processing
+        # completes. Requires after_import (the move fires from the process
+        # job's completion hook) and a destination inside the target's
+        # local archive root; snapshotted now so a mid-chain Settings edit
+        # can't redirect the move.
+        move_target_snapshot, move_err = _validate_after_process_move(
+            after_process_move, after_import, destination,
+        )
+        if move_err is not None:
+            return move_err
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/app.py:21712]: Fixed in c24f975f: `after_import` resolution (including the omitted-default — read from global config for new-workspace requests, since a fresh workspace has no overrides) and the full `after_process_move` validation now run BEFORE `_prepare_import_workspace`, so every 400 on this path — bad shape, unknown target, missing `local_archive_root`, destination outside the root, and the move-without-processing pairing — leaves no stray workspace behind. Tests: `test_after_process_move_failure_does_not_create_workspace` and `test_after_process_move_bad_target_does_not_strand_new_workspace` assert the workspace list and active workspace are unchanged after the 400.

File: vireo/app.py [hunk 3]
```diff
@@ -23097,9 +23313,145 @@ def _apply_no_model_auto_skip(params):
...
+        remote = move_mod.build_remote_move_spec(
+            target, posixpath.dirname(subpath), rsync_bin, ssh_bin)
+        return _start_move_folder_job(
+            runner, workspace_id,
+            folder_id=folder_id,
+            destination=remote["mount_dest_base"],
+            display_dest=move_mod.rsync_dest_spec(
+                target, remote["ssh_dest_base"]),
+            destination_name="",
+            merge=True,
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/app.py:23999]: Verified and fixed in 1c3b6e20. `allow_tracked_merge` is now threaded through `_start_move_folder_job` and the chained enqueue path opts in, so re-importing into an existing shoot folder merges onto the tracked NAS copy via `move_folder`'s exact-overlap reconciliation (already unit-tested in test_move.py) instead of refusing. The pre-copy content-conflict scan still refuses same-name files with differing bytes, a tracked row strictly below the destination is still refused, and the manual move endpoint keeps the default refusal. The happy-path chain test now asserts the flag reaches `move_folder`.

File: vireo/app.py [hunk 4]
```diff
@@ -20349,6 +20434,75 @@ def _validate_after_import(value, db):
...
+            root_real = os.path.realpath(root)
+            inside = os.path.commonpath([dest_real, root_real]) == root_real
+        except ValueError:
+            return None, json_error(
+                "destination is not inside the remote target's local "
+                f"archive root ({root})")
+        if not inside:
+            return None, json_error(
+                "destination is not inside the remote target's local "
+                f"archive root ({root})")
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/app.py]: Fixed in 32d063ac — containment now goes through move.py's `_path_equal_or_descends` (the alias-folding helper the move guards already use, which probes the actual filesystem instead of assuming platform behavior). One subtlety beyond the request-time check: the catalog folders the import creates are spelled like the destination and `minimal_move_set` compares byte-wise at chain time, so accepting a case-alias destination alone would have made the chain later skip every folder as outside_root. When the destination reaches the root only via a case alias, the target snapshot's `local_archive_root` is respelled with the destination's casing. Covered by `test_chain_accepts_case_alias_destination` (skips on case-sensitive filesystems).

File: vireo/config.py [hunk 5]
```diff
@@ -702,6 +704,27 @@ def _coerce_remote_target(entry):
...
+    # the *destination* view of the NAS; the archive root is the local
+    # staging side — pointing it at the mount would "move" files onto
+    # themselves). Invalid values are blanked rather than dropping the
+    # whole target.
+    local_archive_root = (entry.get("local_archive_root") or "").strip()
+    if local_archive_root:
+        if not os.path.isabs(local_archive_root):
+            local_archive_root = ""
+        elif mount_path:
+            try:
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/config.py]: Fixed in 3790d5e3 — the inside-mount containment check now requires an absolute `mount_path` (relative mounts are unusable for transfers anyway), so a valid `local_archive_root` can no longer be blanked depending on the server's launch CWD. Added `test_remote_target_relative_mount_path_keeps_archive_root`.

File: vireo/app.py [hunk 6]
```diff
@@ -23097,9 +23341,153 @@ def _apply_no_model_auto_skip(params):
...
+                    failures.append(f"{entry['subpath']}: {e}")
+            if isinstance(result, dict):
+                if move_ids:
+                    result["move_job_ids"] = move_ids
+                if after_move.get("skip_note"):
+                    # Partial skip: some folders moved but others (e.g.
+                    # photos cataloged on the archive root) stayed local.
+                    result["after_move_note"] = after_move["skip_note"]
+                if failures:
+                    result["after_move_errors"] = failures
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/app.py:24083]: Verified — jobs.html renders only the step tree for stepped jobs (`job.result` shows just in the no-steps fallback, global `job.errors` only when status is failed), so these failures were indeed invisible. Fixed in be8ace0b: added a lock-safe `JobRunner.append_step` and `_chain_after_move` now records a 'Move to NAS' step on the process job — completed with the move count on success, completed-with-warning carrying the honest skip reason when photos stay local, failed with the enqueue errors when nothing started (and on chain-machinery exceptions). Covered by `test_chain_move_enqueue_failure_surfaces_as_step` plus step assertions in the happy-path and root-skip chain tests.

File: vireo/templates/import.html [hunk 7]
```diff
@@ -1222,6 +1244,149 @@ <h3 id="folderBrowserTitle">Select Folder</h3>
...
+// slash. Case-sensitive compare is fine — the server re-validates with
+// realpath/commonpath at enqueue time and stays authoritative.
+function afterMoveNormPath(p) {
+  return String(p || '').replace(/\\/g, '/').replace(/\/+$/, '');
+}
+
+function afterMovePathInsideRoot(destination, root) {
+  const d = afterMoveNormPath(destination);
+  const r = afterMoveNormPath(root);
+  return !!r && (d === r || d.indexOf(r + '/') === 0);
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/templates/import.html:1394]: Fixed in d6b3558f. `afterMovePathInsideRoot` now compares case-insensitively, mirroring the server's alias-folding containment (`_validate_after_process_move` → `_path_equal_or_descends`), so the 'Then move to NAS' row shows for case-alias destinations the backend accepts. `afterMoveNormPath` stays case-preserving since it also renders the preview's NAS-side (case-sensitive) paths and the rel-subpath slicing. The server remains authoritative — on a genuinely case-sensitive filesystem the worst case is now a clear 400 at Start Import rather than a silently missing option (noted in the code comment).

File: vireo/config.py [hunk 8]
```diff
@@ -702,6 +704,32 @@ def _coerce_remote_target(entry):
...
+            # CWD here, making this containment check depend on where the
+            # server happened to be launched — it could blank a perfectly
+            # valid archive root. Relative mounts are unusable for
+            # transfers anyway, so skip the check instead of resolving.
+            try:
+                if os.path.commonpath([
+                    os.path.realpath(local_archive_root),
+                    os.path.realpath(mount_path),
+                ]) == os.path.realpath(mount_path):
+                    local_archive_root = ""
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/config.py]: Verified and fixed in fa8cea18 (pushed by the PR agent; I validated it locally). The inside-mount blanking now routes through `move._path_equal_or_descends` — the same alias-folding helper the move guards and `_validate_after_process_move` use — so a mount/root pair differing only by case on a case-insensitive volume is recognized as the same directory and the archive root is blanked. `move.py` is stdlib-only at module level (no circular import), the different-drives handling is preserved (broadened to OSError/ValueError, preferring to keep a saved config on transient resolution errors), and two regression tests cover the exact-alias and inside-via-alias cases (skipping on case-sensitive filesystems). test_config (74) + test_move_remote (40) + test_jobs_api (208) pass locally with ruff clean.

File: vireo/app.py [hunk 9]
```diff
@@ -20570,6 +20661,91 @@ def _validate_after_import(value, db):
...
+        # "/volumes/photos/…" vs "/Volumes/Photos") is the same directory,
+        # and realpath does not fold case on POSIX — a byte compare would
+        # falsely reject it.
+        from move import _path_equal_or_descends
+        dest_real = os.path.realpath(destination)
+        root_real = os.path.realpath(root)
+        if not _path_equal_or_descends(destination, root):
+            return None, json_error(
+                "destination is not inside the remote target's local "
+                f"archive root ({root})")
```
[the remark this answers is not in the document]
Comment (jss367) [reply vireo/app.py:20847]: Verified and fixed in 3ea7b368. `_validate_after_process_move` now rejects a destination equal to or under the selected target's `mount_path` (alias-folded via `_path_equal_or_descends`, consistent with the root containment check) with an error explaining the import would land directly on the NAS and the chained move would duplicate it under the remote path. A sibling destination inside the archive root remains accepted — both directions covered by `test_after_process_move_rejects_destination_under_mount`.

**Comment by jss367** [comment]
@codex
